worker/jobs/pull/base.py

```diff
@@ -0,0 +1,94 @@
+import ipaddress
+import re
+import requests
+from socket import gethostbyname
+from urllib.parse import urlparse
+
+
+class RemoteFileException(Exception):
+    pass
+
+
+class PullSession(requests.sessions.Session):
+    def __init__(self):
+        super().__init__()
+        self.max_redirects = 5
+        self.headers = {"User-Agent": "Stencila Hub HTTP Client"}
+
+    def get_redirect_target(self, resp):
+        # Override to run each redirect target through the malicious
+        # host check
+        url = super().get_redirect_target(resp)
+        self.check_host(url)
+        return url
+
+    def check_host(self, url):
+        if url is None:
+            return
+
+        url_obj = urlparse(url)
+
+        if url_obj.hostname is None:
+            raise TypeError('Url "{}" appears to have no hostname.'.format(url))
+
+        if is_malicious_host(url_obj.hostname):
+            raise RemoteFileException(
+                "{} is not a valid hostname.".format(url_obj.hostname)
+            )
+
+    def fetch_url(self, url, stream=False):
+        url = convert_raw_content_url(url)
+        self.check_host(url)
+        return self.get(url, stream=stream, allow_redirects=True)
+
+    def pull(self, url, sink):
+        with self.fetch_url(url, stream=True) as response:
+            with open(sink, "wb") as file:
+                for chunk in response:
+                    file.write(chunk)
+
+
+def convert_raw_content_url(url: str) -> str:
```

Comment:
I don't think that we should have this function here. Instead some of this logic is already implemented elsewhere or belongs in forthcoming `pull_github` and `pull_hackmd` functions - ie. functions that pull a specific source type e.g. a `GithubSource`.

The parsing of GitHub URLs is done here:

https://github.com/stencila/hub/blob/03900bacf3e4b814a5c173239a32d71e20a8fed7/director/projects/source_models.py#L338-L360

What the `github_pull` function will have to do is convert a `GithubSource` into a `"https://raw.githubusercontent.com/` URL...

Similarly for HackMD, although we have not yet created a `HackMdSource` model yet.